### DISCUSSION TITLE
Can't connect with unix socket (asyncpg)

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [sqlalchemy/sqlalchemy#8261] [Usage Questions]
The documentation seems to suggest that this should connect with a unix socket:
`"postgresql+asyncpg://{}:{}@/rhymewave_com".format(config["db_user"], config["db_pw"])`

But, when running, I get:
`OSError: Multiple exceptions: [Errno 111] Connect call failed ('::1', 5432, 0, 0), [Errno 111] Connect call failed ('127.0.0.1', 5432)`

While this connects with no problems (although doesn't seem to need the password):
`psql rhymewave_com`

**Comment by zzzeek** [comment]
> The documentation seems to suggest that this should connect with a unix socket: `"postgresql+asyncpg://{}:{}@/rhymewave_com".format(config["db_user"], config["db_pw"])`

our docs are [here](https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#module-sqlalchemy.dialects.postgresql.asyncpg) which have no docs for unix sockets.   It looks like there's one mention of it in asyncpg's docs [here](https://magicstack.github.io/asyncpg/current/api/index.html#connection), though there's no example and it seems kind of vague, so the first step to connecting would be to make a plain asyncpg program that connects in the way you want, which you can then translate into SQLAlchemy's API, most likely by specifying the URL to carefully line up with the expected arguments, or if not, then using [an alternate connectivity approach](https://docs.sqlalchemy.org/en/14/core/engines.html#custom-dbapi-connect-arguments-on-connect-routines).

short answer, you need to experiment with asycnpg first before trying to guess what the SQLAlchemy URL might look like, both to eliminate the asyncpg driver as the issue as well as to get the exact connection arguments needed.

**Reply by Dreamsorcerer** [reply]
> our docs are [here](https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#module-sqlalchemy.dialects.postgresql.asyncpg) which have no docs for unix sockets.

It's further down the page: https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#unix-domain-connections

It only discusses psycopg2, but assumed the same thing would work with asyncpg. The documentation sounds like you are special-casing psycopg2 and sending the host as None, whereas asyncpg, it seems you are just passing the dsn only.

It seems that I can do `postgresql:///rhymewave_com?host=/var/run/postgresql`, but it then fails to find the file for some reason. It should also pick up the default unix socket if no host is provided according to those docs, but asyncpg actually tries to connect to localhost in that situation. Will follow up with asyncpg then, there's no issue with sqlalchemy, it sounds like the sqlalchemy examples _should_ work with asyncpg.

**Reply by zzzeek** [reply]
> > our docs are [here](https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#module-sqlalchemy.dialects.postgresql.asyncpg) which have no docs for unix sockets.
> 
> It's further down the page: https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#unix-domain-connections
> 
> It only discusses psycopg2, but assumed the same thing would work with asyncpg. 

oh no, not at all.  that's the psycopg2 driver doc.  not related in any way to asyncpg

> The documentation sounds like you are special-casing psycopg2 and sending the host as None, whereas asyncpg, it seems you are just passing the dsn only.

absolutely, every driver is completely different and has entirely different connectivity approaches.   psycopg2 uses libpq directly.   asyncpg does not.  totally different worlds.   pg8000, yet another one.    the docs are broken into driver-specific sections like that for this reason.


> 
> It seems that I can do `postgresql:///rhymewave_com?host=/var/run/postgresql`, but it then fails to find the file for some reason. It should also pick up the default unix socket if no host is provided according to those docs, but asyncpg actually tries to connect to localhost in that situation. Will follow up with asyncpg then, there's no issue with sqlalchemy, it sounds like the sqlalchemy examples _should_ work with asyncpg.

OK but step one is to make a plain asyncpg test case first.  because if they *do* support that, we just need to know the format and I can see if the URL can do that now.

**Reply by Dreamsorcerer** [reply]
> oh no, not at all. that's the psycopg2 driver doc. not related in any way to asyncpg#

Oh, I see now. I was just searching the page for unix sockets. I didn't look at any of the headers or anything above that.

**Reply by Dreamsorcerer** [reply]
OK, finally got somewhere. The problem basically seems to have been Debian using a different default port. Seems I have a connection working now with: `"postgresql+asyncpg:///rhymewave_com?port=5433"`

Although, curiously, that fails when used directly in asyncpg itself, as it seems to default to 5 hosts, and therefore requires either `"postgresql:///rhymewave_com?host=/var/run/postgresql&port=5433"` or `"postgresql:///rhymewave_com?port=5433,5433,5433,5433,5433"`.

**Reply by zzzeek** [reply]
we are only using asyncpg so that URL is passing in some working form of argument to the driver
